provider/image.go

```diff
@@ -0,0 +1,314 @@
+	img := Image{
+		Name:     inputs["imageName"].StringValue(),
+		SkipPush: inputs["skipPush"].BoolValue(),
+		Registry: reg,
+	}
+
+	build := marshalBuild(inputs["build"])
+	cache := getCachedImages(img, inputs["build"])
+
+	build.CachedImages = cache
+	img.Build = build
+
+	docker, err := client.NewClientWithOpts(client.FromEnv)
+
+	if err != nil {
+		panic(err)
+	}
+
+	err = p.host.Log(ctx, "info", urn, "Building the image")
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	// make the build context
+	tar, err := archive.TarWithOptions(img.Build.Context, &archive.TarOptions{})
+	if err != nil {
+		return "", nil, err
+	}
+
+	// make the build options
+
+	opts := types.ImageBuildOptions{
+		Dockerfile: img.Build.Dockerfile,
+		Tags:       []string{img.Name}, //this should build the image locally, sans registry info
+		Remove:     true,
+		//CacheFrom:  img.Build.CachedImages, // TODO: this needs a login, so needs to be handled differently.
+		BuildArgs: build.Args,
+		//Version:   types.BuilderBuildKit, // TODO: parse this setting from the `env` input
+	}
+
+	imgBuildResp, err := docker.ImageBuild(context.Background(), tar, opts)
+	if err != nil {
+		return "", nil, err
+	}
+
+	defer imgBuildResp.Body.Close()
+	// Print build logs to terminal
+	scanner := bufio.NewScanner(imgBuildResp.Body)
+	for scanner.Scan() {
+		err := p.host.Log(ctx, "info", urn, scanner.Text())
+		if err != nil {
+			return "", nil, err
+		}
+	}
+
+	// if we are not pushing to the registry, we return after building the local image.
+	if img.SkipPush {
+		outputs := map[string]interface{}{
+			"dockerfile":     img.Build.Dockerfile,
+			"context":        img.Build.Context,
+			"registryServer": img.Registry.Server,
+		}
+
+		pbstruct, err := plugin.MarshalProperties(
+			resource.NewPropertyMapFromMap(outputs),
+			plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true},
+		)
+		return img.Name, pbstruct, err
+	}
+
+	err = imgBuildResp.Body.Close()
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	err = p.host.Log(ctx, "info", urn, "Pushing Image to the registry")
+
+	if err != nil {
+		return "", nil, err
+	}
+	// Quick and dirty auth; we can also preconfigure the client itself I believe
+
+	var authConfig = types.AuthConfig{
+		Username:      img.Registry.Username,
+		Password:      img.Registry.Password,
+		ServerAddress: img.Registry.Server,
+	}
+
+	authConfigBytes, err := json.Marshal(authConfig)
+
+	if err != nil {
+		return "", nil, errors.Wrap(err, "Error parsing authConfig")
+	}
+	authConfigEncoded := base64.URLEncoding.EncodeToString(authConfigBytes)
+
+	pushOpts := types.ImagePushOptions{RegistryAuth: authConfigEncoded}
+
+	// By default, we push our image with the qualified image name from the input, without extra tagging.
+	pushOutput, err := docker.ImagePush(context.Background(), img.Name, pushOpts)
+
+	if err != nil {
+		return "", nil, err
+	}
+
+	defer pushOutput.Close()
+
+	// Print push logs to terminal
+	pushScanner := bufio.NewScanner(pushOutput)
```

Comment:
Excellent! Scanning semi structured output can be hard to get right. It could help us to lift this scanning piece into its own func if we need to test it a bit.